src/transactions.rs

```diff
@@ -0,0 +1,456 @@
+    }
+    pub fn inputs(&self) -> &Vec<RecordOpening> {
+        &self.inputs
+    }
+    pub fn outputs(&self) -> &Vec<RecordOpening> {
+        &self.outputs
+    }
+    pub fn time(&self) -> &DateTime<Local> {
+        &self.time
+    }
+    pub fn asset(&self) -> &AssetCode {
+        &self.asset
+    }
+    pub fn kind(&self) -> &TransactionKind<L> {
+        &self.kind
+    }
+    pub fn senders(&self) -> &Vec<UserAddress> {
+        &self.senders
+    }
+    pub fn receivers(&self) -> &Vec<(UserAddress, RecordAmount)> {
+        &self.receivers
+    }
+    pub fn fee_change(&self) -> &Option<RecordAmount> {
+        &self.fee_change
+    }
+    pub fn asset_change(&self) -> &Option<RecordAmount> {
+        &self.asset_change
+    }
+    pub fn receipt(&self) -> &Option<TransactionReceipt<L>> {
+        &self.receipt
+    }
+}
+
+type TransactionsStore<L> = KeyValueStore<TransactionUID<L>, Transaction<L>>;
+
+/// An editor to create or update the transaction or transactions store.
+pub struct TransactionEditor<'a, L: Ledger + Serialize + DeserializeOwned> {
+    transaction: Transaction<L>,
+    store: &'a mut TransactionsStore<L>,
+}
+
+impl<'a, L: Ledger + Serialize + DeserializeOwned> TransactionEditor<'a, L> {
+    /// Create a transaction editor.
+    fn new(store: &'a mut TransactionsStore<L>, transaction: Transaction<L>) -> Self {
+        Self { transaction, store }
+    }
+
+    /// Add fee change record to the transaction once it is certain
+    pub fn with_fee_change(mut self, amount: RecordAmount) -> Self {
+        self.transaction.fee_change = Some(amount);
+        self
+    }
+
+    /// Add asset change record to the transaction once it is certain
+    pub fn with_asset_change(mut self, amount: RecordAmount) -> Self {
+        self.transaction.asset_change = Some(amount);
+        self
+    }
+
+    /// Add the transaction receipt when it is recieved
+    pub fn with_receipt(mut self, receipt: TransactionReceipt<L>) -> Self {
+        self.transaction.receipt = Some(receipt);
+        self
+    }
+
+    /// Add the transaction hash, should be called after this transaction is committed
+    pub fn with_hash(mut self, hash: TransactionHash<L>) -> Self {
+        self.transaction.hash = Some(hash);
+        self
+    }
+
+    /// Add the UIDs of memos we are waiting on to complete the transaction
+    pub fn add_pending_uids(mut self, uids: &[u64]) -> Self {
+        for uid in uids {
+            self.transaction.pending_uids.insert(*uid);
+        }
+        self
+    }
+    /// remove a UID of a memo we were waiting because it was received
+    pub fn remove_pending_uid(mut self, uid: u64) -> Self {
+        self.transaction.pending_uids.remove(&uid);
+        self
+    }
+
+    /// Save the transaction to the store.
+    ///
+    /// Returns the stored transaction.
+    pub fn save(&mut self) -> Result<Transaction<L>, KeystoreError<L>> {
+        self.store.store(&self.transaction.uid, &self.transaction)?;
```

Comment:
Since our secondary keys are mutable (i.e. `with_hash`, `add_pending_uids`, `remove_pending_uid`) I think we need to update the indices when we save here. Easiest way is to just reload from disk, otherwise we might need to keep track of changes made to the editor.